[RDY] fix SIGTERM/SIGHUP for jobs

blueyed opened the pull request:
TODO:
- [x] test for the SIGKILL handler to kick in (e.g. by trapping/masking the
  TERM signal)
- [x] revisit docs

Ref: https://github.com/neovim/neovim/pull/6820

blueyed committed:
jobstop/process_stop: send SIGTERM directly

This reverts the revert of #6644 (7c1a5d1d4), and handles it properly
now (with tests).

blueyed commented on src/nvim/event/process.c:
```diff
@@ -242,7 +241,7 @@ void process_stop(Process *proc) FUNC_ATTR_NONNULL_ALL
     case kProcessTypePty:
       // close all streams for pty processes to send SIGHUP to the process
```
Not sure if this comment is still correct anymore, but the `process_close_streams` call is needed.

oni-link commented on src/nvim/event/process.c:
```diff
@@ -253,7 +254,8 @@ void process_stop(Process *proc) FUNC_ATTR_NONNULL_ALL
     // will periodically check if a signal should be send to a to the job
     DLOG("Starting job kill timer");
-    uv_timer_start(&loop->children_kill_timer, children_kill_cb, 100, 100);
+    uv_timer_start(&loop->children_kill_timer, children_kill_cb, KILL_TIMEOUT,
```
The timeout is in milliseconds not nanoseconds.

blueyed commented on test/functional/autocmd/termclose_spec.lua:
```diff
@@ -14,13 +14,38 @@ describe('TermClose event', function()
+    command('sleep 200m')
+    -- command('terminal')
+    -- command('call jobsend(b:terminal_job_id, ["trap \'\' HUP", "(echo $$; trap;) >> ~/.nvimlog", "sleep 5"])')
+    -- eq(2, eval('b:terminal_job_pid'))
```
Any comments about this?
I would like to avoid the sleep here, but ensure that the terminal is up and running (having setup the trap).
/cc @justinmk

blueyed replied:
I can also see that the additional tests do not really fit the scope of only testing the autocommand after all.
Any suggestions?

justinmk replied:
> the additional tests do not really fit the scope of only testing the autocommand after all

Don't worry about that, I'll shuffle it around before merging.

blueyed replied:
Thanks.
Any idea how to avoid the sleep here?

justinmk commented on src/nvim/event/process.c:
```diff
@@ -21,9 +21,12 @@
+#define KILL_TIMEOUT_NS (KILL_TIMEOUT_MS * 1000000)
+// Time for a pty process to exit due to SIGHUP before we send TERM.
+#define PTY_TERM_TIMEOUT_MS 1000
+#define PTY_TERM_TIMEOUT_NS (PTY_TERM_TIMEOUT_MS * 1000000)
```
Why use different timeouts for pty? And why send sigterm, just skip to sigkill if sighup didn't work.

blueyed replied:
Good point.  But it would cause a 2s delay delay (only for a constructed/non-default case though).

justinmk replied:
2s vs 1s doesn't really matter, at least I can't think of a reason why it should be different for pty vs non-pty jobs.

jamessan replied:
Why _would_ you skip SIGTERM? It's not uncommon for SIGHUP to have special handling other than tearing down the process.

justinmk replied:
@jamessan If SIGHUP fails is SIGTERM needed? And is SIGTERM appropriate for pty processes?

blueyed replied:
SIGHUP is used with pty processes by default (without timeout), and SIGTERM with others (without timeout).

If SIGHUP is not enough (as I understood from @jamessan), then SIGTERM gets used after 1s for pty processes (which could also be 2s though).  It should not send SIGKILL though directly, but only after another timeout.
I think having a single value for the timeout (of 2s) could simplify it a bit, which results in 4s for a non-responsive pty process.  Two different macros/values/names for it make it more descriptive though, and a shorter timeout for SIGTERM can be used then.

justinmk replied:
Avoiding extra logic and variables is the priority. The time out happens in unusual cases, doesn't matter if it takes 10 seconds.

blueyed replied:
Ok. There is only one macro now, but it needs extra logic for pty jobs before sending the SIGKILL.

blueyed commented on src/nvim/event/process.c:
```diff
@@ -269,15 +273,18 @@ static void children_kill_cb(uv_timer_t *handle)
-    } else if (elapsed >= KILL_TIMEOUT) {
-      ILOG("Sending SIGKILL to pid %d", proc->pid);
-      uv_kill(proc->pid, SIGKILL);
+    uint64_t elapsed = (now - proc->stopped_time) / 1000000 + 1;
```
The `+1` here is ugly: elapsed is typically 1999 or 3999 here.
I've tried to use `uv_update_time(&loop->uv);` when starting the timer to no avail.

justinmk replied:
Was it waiting 2s longer, before this change?

blueyed replied:
Yeah.. the condition was always 1ms short.

blueyed commented on test/functional/autocmd/termclose_spec.lua:
```diff
@@ -14,13 +14,51 @@ describe('TermClose event', function()
+    nvim('set_option', 'shell', 'sh')
+    nvim('set_option', 'shellcmdflag', '-c')
+    command([[
+      let g:test_job = jobstart('trap "" HUP TERM && echo 1 && sleep 60', {'pty': 1, 'on_stdout': {-> execute('let g:test_job_started = 1')}, 'on_exit': {-> execute('let g:test_job_exited = 1')}})
```
I've not managed to wrap this into multiple line.. tried several levels of backslashes in front of the continuation lines.

justinmk replied:
It's vimscript, so just one `\` at the start of the line will work.

justinmk replied:
Because command() is a single command. For line continuations to work one needs `helpers.source()`

justinmk replied:
Anyways this would be best:

```lua
command([[ ... ]]
  .. [[ ... ]])
```

oni-link commented on src/nvim/event/process.c:
```diff
@@ -269,15 +273,18 @@ static void children_kill_cb(uv_timer_t *handle)
+        ILOG("Sending SIGTERM to pid %d", proc->pid);
+        uv_kill(proc->pid, SIGTERM);
+        proc->term_sent = true;
+      } else if (elapsed >= KILL_TIMEOUT_MS * (proc->type == kProcessTypePty)) {
```
If this condition is checked, it is always true.

blueyed replied:
Not necessarily, if there are several pty jobs waiting to be killed: the `repeat` for the timer is `KILL_TIMEOUT_MS`, but if new jobs are going to be queued while the timer has not been removed, this might interfere - or at least that is what I was thinking.

oni-link replied:
We have already checked in an outer `if` that the condition `elapsed >= KILL_TIMEOUT_MS` is true, so the check here is also true ( `elapsed >= KILL_TIMEOUT_MS * {0/1}`).

blueyed replied:
Oh.. it should actually be {1,2} here..!
But it can be clarified on the outer loop indeed.

blueyed replied:
Fixed.

jamessan commented:
Are the lint fixes and the doc updates the only things left on this?

oni-link commented on src/nvim/event/process.c:
```diff
@@ -269,15 +273,18 @@ static void children_kill_cb(uv_timer_t *handle)
+        ILOG("Sending SIGKILL to pid %d", proc->pid);
+        uv_kill(proc->pid, SIGKILL);
+      }
     }
```
We could remove the member `proc->term_sent` with something like this:
```c
    if (elapsed >= KILL_TIMEOUT_MS) {
      int sig = proc->type == kProcessTypePty && elapsed < KILL_TIMEOUT_MS * 2
                    ? SIGTERM
                    : SIGKILL;
      ILOG("Sending %s to pid %d", sig == SIGTERM ? "SIGTERM" : "SIGKILL",
           proc->pid);
      uv_kill(proc->pid, sig);
    }
```

blueyed replied:
Thanks, taken!

blueyed commented:
- Removed term_sent, thanks to @oni-link.
- Fixed formatting in tests.

I've thought there was still some sleep somewhere, but it was removed already.

So I think only revisiting the docs is left here, and maybe if there is a way to not cause the tests slowing down by 2 + 4 seconds for the TERM/KILL tests?!
Can we somehow advance/fake the current time during tests?

oni-link commented on src/nvim/event/process.c:
```diff
@@ -126,7 +126,6 @@ void process_teardown(Loop *loop) FUNC_ATTR_NONNULL_ALL
       CREATE_EVENT(loop->events, process_close_handles, 1, proc);
     } else {
       uv_kill(proc->pid, SIGTERM);
```
This could also be removed, because `process_stop()` sends signal `SIGTERM` without waiting.

blueyed replied:
Removed.

justinmk commented:
I just noticed a "uv_loop_close hang" detected by https://github.com/neovim/neovim/pull/6827 on [travis master](https://s3.amazonaws.com/archive.travis-ci.org/jobs/241838820/log.txt?X-Amz-Expires=30&X-Amz-Date=20170613T223324Z&X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=AKIAJRYRXRSVGNKPKO5A/20170613/us-east-1/s3/aws4_request&X-Amz-SignedHeaders=host&X-Amz-Signature=178bead7067de95c334dbcbb82e8bec79f93b5db50db16444648f483e36bc6cc). Will be interested to see if that continues after this PR.

```
[32m[2m[----------][0m[0m Running tests from [36m/home/travis/build/neovim/neovim/test/functional/autocmd/termclose_spec.lua[0m
[32m[2m[ RUN      ][0m[0m TermClose event triggers when terminal job ends: -- Output to stderr:
Vim: Caught deadly signal 'SIGHUP'

Vim: Finished.

CMake Error at /home/travis/build/neovim/neovim/cmake/RunTests.cmake:50 (message):
  Running functional tests failed with error: 1.

===============================================================================
NVIM_LOG_FILE: /home/travis/build/neovim/neovim/build/.nvimlog
2017/06/11 23:17:08 ERROR 9995/loop_close:100: uv_loop_close() hang?
[--I] signal   0x9ff948
[-AI] async    0x9ff790
[R--] signal   0x9f93a0

travis_fold:end:-tests-functionaltests
[0KSuite /tests/functionaltests failed, summary:
```

blueyed commented:
@justinmk 
This PR includes #6827 already (I've merged master in here).

justinmk commented:
Restarted the build just for fun (wondering if build timeouts become less frequent with this PR). Also do you plan to squash it up?

blueyed commented:
> Also do you plan to squash it up?

Of course.
After revisiting docs, and making the tests faster.  Do you have an idea about that? (see my previous comment)

blueyed commented:
Rebased.

@justinmk 
You've said that you would move the tests around when merging.
Should I squash it already before?

The docs seem to be fitting still:

```
jobstop({job})						{Nvim} *jobstop()*
		Stop a job created with |jobstart()| by sending a `SIGTERM`
		to the corresponding process. If the process doesn't exit
		cleanly soon, a `SIGKILL` will be sent. When the job is
		finally closed, the exit handler provided to |jobstart()| or
		|termopen()| will be run.
		See |job-control| for more information.
```

justinmk commented:
yes, please squash as you see fit.

blueyed commented:
Done.

justinmk commented:
still occasionally seeing a hang provoked by `termclose_spec.lua`. e.g. https://travis-ci.org/neovim/neovim/jobs/263860335

```
NVIM_LOG_FILE: /home/travis/build/neovim/neovim/build/.nvimlog
2017/08/12 18:26:26 ERROR 12455/loop_close:100: uv_loop_close() hang?
[--I] signal   0xa1b9e8
[-AI] async    0xa1b830
[R--] signal   0xa153c0
```
